main.tf

```diff
@@ -17,6 +18,22 @@ module "autoscaler" {
   worker_pool_id            = var.worker_pool_id
 }
 
+module "lifecycle_manager" {
+  count  = local.autoscaling_enabled && length(var.instance_refresh) > 0 ? 1 : 0
+  source = "./lifecycle_manager"
+
+  additional_tags            = var.additional_tags
+  api_key_ssm_parameter_arn  = module.autoscaler[0].ssm_parameter_arn
+  api_key_ssm_parameter_name = module.autoscaler[0].ssm_parameter_name
+  auto_scaling_group_arn     = module.asg.autoscaling_group_arn
+  autoscaling_configuration  = var.autoscaling_configuration
+  aws_partition_dns_suffix   = data.aws_partition.current.dns_suffix
+  aws_region                 = data.aws_region.this.name
+  base_name                  = local.base_name
+  iam_permissions_boundary   = var.iam_permissions_boundary
+  worker_pool_id             = var.worker_pool_id
+}
```

Comment:
The only thing I dont really like about this change is that the lifecycle hook is dependent on the autoscaler being enabled. I did this because we need an api id and key from spacelift and its passed into the autoscaler already. So to prevent folks from having to pass it into the module multiple times, I just required autoscaling to be enabled for this to work.

Our options are:
1) instance refreshing only supports graceful shutdowns if autoscaling is enabled
2) we introduce a breaking change to the module and move the api config out of the autoscaling config variable
3) we have the users pass in the api key more than once.

Reply:
Does the lifecycle hook _depend_ on the autoscaler at all? If not, I'd argue that option 2 is the best. Sure, it's a breaking change, but we generally shouldn't be scared of breaking changes; we just need to make sure they're communicated properly (e.g. as they were for 3.0.0).

I'd also suggest that the lifecycle hook be behind a feature-gate (just like autoscaling is). That way, if the user doesn't want to provide an API key at all, they can disable _both_ features. As for what it's default is (enabled/disabled), I could go either way. It makes sense to encourage users to have the lifecycle hooks enabled, but if they require API keys, then it _may_ be best to leave it disabled by default? But TBH, I can't imagine anyone would want to _not_ have the lifecycle hooks.

Reply:
Agreed, I will be turning this into a breaking change. We will just need to communicate it out before release.